lib/experimental/modules/class-gutenberg-modules.php

```diff
@@ -20,64 +20,90 @@ class Gutenberg_Modules {
 	 *
-	 * @var string[]
+	 * @var array
 	 */
-	private static $enqueued = array();
+	private static $enqueued_modules_before_register = array();
 
 	/**
-	 * Registers the module if no module with that module identifier already
-	 * exists.
+	 * Registers the module if no module with that module identifier has already
+	 * been registered.
 	 *
-	 * @param string           $module_identifier The identifier of the module. Should be unique. It will be used in the final import map.
-	 * @param string           $src               Full URL of the module, or path of the script relative to the WordPress root directory.
-	 * @param array            $dependencies      Optional. An array of module identifiers of the static and dynamic dependencies of this module. It can be an indexed array, in which case all the dependencies are static, or it can be an associative array, in which case it has to contain the keys `static` and `dynamic`.
-	 * @param string|bool|null $version           Optional. String specifying module version number. It is added to the URL as a query string for cache busting purposes. If SCRIPT_DEBUG is true, a timestamp is used. If it is set to false, a version number is automatically added equal to current installed WordPress version. If set to null, no version is added.
+	 * @param string            $module_identifier The identifier of the module. Should be unique. It will be used in the final import map.
+	 * @param string            $src               Full URL of the module, or path of the script relative to the WordPress root directory.
+	 * @param array             $dependencies      Optional. An array of module identifiers of the dependencies of this module. The dependencies can be strings or arrays. If they are arrays, they need an `id` key with the module identifier, and can contain a `type` key with either `static` or `dynamic`. By default, dependencies that don't contain a type are considered static.
+	 * @param string|false|null $version           Optional. String specifying module version number. Defaults to false. It is added to the URL as a query string for cache busting purposes. If SCRIPT_DEBUG is true, the version is the current timestamp. If $version is set to false, the version number is the currently installed WordPress version. If $version is set to null, no version is added.
 	 */
 	public static function register( $module_identifier, $src, $dependencies = array(), $version = false ) {
-		// Register the module if it's not already registered.
 		if ( ! isset( self::$registered[ $module_identifier ] ) ) {
-			$deps = array(
-				'static'  => isset( $dependencies['static'] ) || isset( $dependencies['dynamic'] ) ? $dependencies['static'] ?? array() : $dependencies,
-				'dynamic' => isset( $dependencies['dynamic'] ) ? $dependencies['dynamic'] : array(),
-			);
+			$deps = array();
+			foreach ( $dependencies as $dependency ) {
+				if ( is_array( $dependency ) && isset( $dependency['id'] ) ) {
```

Comment:
`isset()` is a language construct and it's safe to run this as a single condition

```php
if ( isset( $dependency['id'] ) ) {
```

even if `$dependency` were `null` it wouldn't throw. the `is_array()` check of course is nice, indicating we want it to be an array, but also, if `['id']` is set then it has to be array-like anyway

Reply:
Nice! Done in https://github.com/WordPress/gutenberg/pull/57231/commits/e6b510e3715f7ebfa94e00dc444b6a9a040e0ca9.